File: src/com/mdgeorge/algebra/concept/GroupHom.java
package com.mdgeorge.algebra.concept;

import com.mdgeorge.algebra.properties.CommutesWith;

/**
 * A group homomorphism h from a group D (the domain) to a group C (the codomain)
 * is a function that commutes with the group operation on C and D.  In other
 * words, for all a, b in D, h(a + b) = h(a) + h(b).
 *
 * @see <a href="http://en.wikipedia.org/wiki/Group_homomorphism">Group homomorphism on Wikipedia</a>
 * @author mdgeorge
 */
public interface GroupHom <DE, CE>
         extends SetHom   <DE, CE>
{
	@Override
	@CommutesWith("com.mdgeorge.algebra.concept.Group.plus")
	CE ap (DE e);
	
	@Override Group<DE> domain();
	@Override Group<CE> codomain();
}

File: src/co/ndan/algebra/concept/GroupHom.kt
package co.ndan.algebra.concept

import co.ndan.algebra.properties.CommutesWith

/*
 * A group homomorphism h from a group D (the domain) to a group C (the codomain)
 * is a function that commutes with the group operation on C and D.  In other
 * words, for all a, b in D, h(a + b) = h(a) + h(b).
 *
 * @see [Group homomorphism on Wikipedia](http://en.wikipedia.org/wiki/Group_homomorphism)
 *
 *
 */
interface GroupHom<DE, CE> : SetHom<DE, CE> {
  @CommutesWith("co.ndan.algebra.concept.Group.plus")
  override fun ap(e: DE): CE

  override fun domain(): Group<DE>
  override fun codomain(): Group<CE>
}
